connmgr/connmanager_test.go

```diff
@@ -454,6 +455,77 @@ func TestNetworkFailure(t *testing.T) {
 	}
 }
 
+// TestMultipleFailedConns ensures that the connection manager remains
+// responsive when there are multiple simultaneous failed connections for
+// persistent peers in the retry state.
+func TestMultipleFailedConns(t *testing.T) {
+	// Override the max retry duration for this test since it relies on having
+	// multiple connections in the retry state.
+	curMaxRetryDuration := maxRetryDuration
+	maxRetryDuration = 500 * time.Millisecond
+	defer func() {
+		maxRetryDuration = curMaxRetryDuration
+	}()
+
+	var closeOnce sync.Once
+	const targetFailed = 5
+	var dials uint32
+	reachedTargetFailed := make(chan struct{})
+	errDialer := func(ctx context.Context, network, addr string) (net.Conn, error) {
+		totalDials := atomic.AddUint32(&dials, 1)
+		if totalDials >= targetFailed {
+			closeOnce.Do(func() { close(reachedTargetFailed) })
+		}
+		return nil, errors.New("network down")
+	}
+	cmgr, err := New(&Config{
+		RetryDuration: maxRetryDuration,
+		Dial:          errDialer,
+	})
+	if err != nil {
+		t.Fatalf("New error: %v", err)
+	}
+	ctx, shutdown, wg := runConnMgrAsync(context.Background(), cmgr)
+
+	// Establish several connection requests to localhost IPs.
+	for i := 0; i < targetFailed; i++ {
+		cr := &ConnReq{
+			Addr: &net.TCPAddr{
+				IP:   net.ParseIP(fmt.Sprintf("127.0.0.%d", i)),
+				Port: 18555,
+			},
+			Permanent: true,
+		}
+		go cmgr.Connect(ctx, cr)
+	}
+
+	// Wait for the target number of dials and ensure they happen simultaneously
+	// by checking it happens before the retry timeout.
+	select {
+	case <-reachedTargetFailed:
+	case <-time.After(20 * time.Millisecond):
```

Comment:
I wouldn't be surprised if the CI VMs hit the occasionally as the github actions jobs seem to get preempted for silly long periods fairly frequently.

Reply:
Yeah, it's possible for sure.  I was worried about that when writing and updating a lot of the tests which also rely on various timeouts, but there are quite a lot of them that use these timeouts and I haven't seen any false positives yet.

We might ultimately need to increase them all a bit to something more generous, even though the timeouts are already rather generous since the event it's waiting on happens in a less than 1ms.